testbed/HIL_Firmware/common/dac.c

```diff
@@ -0,0 +1,135 @@
+#include <stdio.h>
+#include "dac.h"
+#include "freertos/FreeRTOS.h"
+#include "esp_err.h"
+#include "driver/twai.h"
+#include "driver/spi_master.h"
+#include "driver/dac_oneshot.h"
+
+//retain 4 least significant bits
+#define V_OUT_MASK 0x0F;
+
+twai_message_t message_status = {
+    .identifier = MESSAGE_STATUS,
+    .extd = 1,
+    .data_length_code = 1,
+};
+
+//convert V_REF from mV to V
+const double STEPV12 = (V_REF/1000)/MAX_STEPS_12;
+const double STEPV8 = (V_REF/1000)/MAX_STEPS_8;
+
+static bool channel1=false;
+
+static dac_oneshot_handle_t chan0_handle;
+static dac_oneshot_config_t chan0_cfg = {
+    .chan_id = DAC_CHAN_0,
+};
+
+int setDacVoltage(float voltage)
+{
+    esp_err_t fault = ESP_OK;
+
+    //above Vref, set to Vref, below 0 set to 0 
+    if(voltage > V_REF || voltage < 0.0f)
+    {
+        if(voltage > V_REF)
+        {
+            printf("voltage too high\n");
+            voltage = V_REF;
+        }
+        else
+        {
+            voltage = 0.0f;
+        }
+    }
+
+    uint8_t Vout = (voltage/1000)/STEPV8;
+    
+    //checking if channel already initialized
+    if(channel1 == false)
+    {
+        dac_oneshot_new_channel(&chan0_cfg, &chan0_handle);
+        channel1 = true;
+    }
+    
+    fault = dac_oneshot_output_voltage(chan0_handle, Vout);
+    if(fault == ESP_OK)
+    {
+        message_status.data[0] = 1;
+        twai_transmit(&message_status, portMAX_DELAY);
+        printf("setting voltage to %fV in channel 1\n", voltage); 
+
+        return ESP_OK;  
```

Comment:
Always use if statements for failing cases so its easier for the next person to add more. This passing case should be at the end of the function and move the failing case here